src/esw/imu_driver.py

```diff
@@ -1,20 +1,94 @@
 #!/usr/bin/env python3
+import numpy as np
 import rospy
 from sensor_msgs.msg import Temperature, Imu, MagneticField
-from geometry_msgs.msg import Quaternion, Vector3
+from geometry_msgs.msg import Quaternion, Vector3, PoseWithCovarianceStamped, PoseWithCovariance, Vector3Stamped, Pose
 from std_msgs.msg import Header
 from mrover.msg import CalibrationStatus, ImuAndMag
+from tf.transformations import quaternion_about_axis, quaternion_multiply, rotation_matrix, quaternion_from_matrix
+from typing import Tuple
+from copy import deepcopy
 
 import serial
 from serial import SerialException, SerialTimeoutException
 
 
+def get_covariances() -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Get the IMU covariances for orientation, gyroscope, accelerometer, and magnetometer pose
+    from rosparam.
+
+    :returns: a tuple containing the orientation, gyro, accel, and mag pose covariance matrices in order
+    """
+    orientation = np.array(rospy.get_param("global_ekf/imu_orientation_covariance"))
+    gyro = np.array(rospy.get_param("global_ekf/imu_gyro_covariance"))
+    accel = np.array(rospy.get_param("global_ekf/imu_accel_covariance"))
+    mag_pose = np.array(rospy.get_param("global_ekf/imu_mag_pose_covariance"))
+    return (orientation, gyro, accel, mag_pose)
+
+
+def publish_mag_pose(pub: rospy.Publisher, msg: Vector3Stamped, covariance: np.ndarray, frame: str):
+    """
+    Estimates the rover's yaw angle from the magnetometer measurement and publishes it
+    as a PoseWithCovarianceStamped for use in the EKF.
+
+    :param pub: PoseWithCovarianceStamped publisher used to publish the message
+    :param msg: magnetic field message used to estimate the yaw angle
+    :param covariance: 6x6 covariance matrix for the estimated pose,
+                       each row/col is [x, y, z, roll, pitch, yaw]
+    :param frame: the frame in which to publish the estimated pose
+    """
+
+    # get unit magnetic field vector in the XY plane
+    mag_vec = np.array([msg.vector.x, msg.vector.y])
+    mag_vec = mag_vec / np.linalg.norm(mag_vec)
+
+    # convert it to a rotation about the Z axis
+    rotationMatrix = np.array(
+        [[mag_vec[1], -1 * mag_vec[0], 0, 0], [mag_vec[0], mag_vec[1], 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]
+    )
+    q = quaternion_from_matrix(rotationMatrix)
+
+    # publish as a pose with the configured mag covariance matrix for rotation
+    pub.publish(
+        PoseWithCovarianceStamped(
+            header=Header(stamp=msg.header.stamp, frame_id=frame),
+            pose=PoseWithCovariance(pose=Pose(orientation=Quaternion(*q)), covariance=covariance.flatten().tolist()),
```

Comment:
Why is this message a PoseWithCovariance when its only a heading?

Reply:
Because the EKF node only accepts Imu, PoseWithCovarianceStamped, TwistWithCovarianceStamped, and Odometry messages, so I figured pose was probably the best one

Reply:
how does it know to ignore the position part then? Do you tell it that?